downloader/src/main/kotlin/vcs/GitWorkingTree.kt

```diff
@@ -70,18 +70,16 @@ open class GitWorkingTree(
             SubmoduleWalk.forIndex(repo).use { walk ->
                 while (walk.next()) {
                     val path = "$prefix${walk.path}"
-                    paths += path
 
                     if (walk.repository == null) {
                         logger.info {
                             "Git submodule at '$path' not initialized. Cannot recursively list its submodules."
                         }
-
-                        continue
-                    }
-
-                    walk.repository.use { submoduleRepo ->
-                        listSubmodules(path, submoduleRepo, paths)
+                    } else {
```

Comment:
commit: I'm not sure this change is ok:

1. When the working tree is used for the sources of a project, there could be a point in saying that non-recursive cloning is fine. Or maybe the source have been clone manually, and only some submodules are initialized.
2. When the working tree is used for the sources of a dependency, I believe ORT does recursive cloning always. Also the provenance resolver recursively determines submodules. 

Is this fix targeting an issue which appears with (1) projects or with (2) dependencies?

Reply:
it targets issues with (2) dependencies: these appear both in git projects, which depend on protobuf, as well as in monorepo projects which contain the protobuf code in a subfolder.

Reply:
@bennati : Just to double check: By dependencies, we're talking about packages consumed via a package manager (not dependencies included in the source code of the project like in C++ / Git-Repo projects), right?

Reply:
both: in the case of the git project, protobuf is fetched using a package manager, in the case of the monorepo project, protobuf is part of the source code. The problem in both cases comes from uninitialized submodules inside protobuf's code.

Reply:
> If you see another behavior @bennati, then we have yet another problem.

But independently of that, I still believe that not listing something that's not there is the right thing to do in this context, no matter *why* it's not there.

Reply:
ORT downloader does clone recursively, but that does not mean it always resolves all submodules correctly: for example we had a repository which contained a submodule pointing to itself, which caused the scanner to crash. We fixed it by removing the recursive submodule in that case.

Reply:
> that does not mean it always resolves all submodules correctly

Well, the downloader *should* resolve all submodules correctly *if* they are properly configured.

> we had a repository which contained a submodule pointing to itself

Do you think it makes sense add a check fort this, and skip such "submodules", too?

Reply:
> But independently of that, I still believe that not listing something that's not there is the right thing to do in this context, no matter _why_ it's not there.

Hm, as we ditched the legacy scanner now all scanning of dependencies uses the provenance resolver.
An implementation detail of the provenance resolver is that it's using Git's recursive submodule initialization.
To my understanding this algorithm is sub-optimal and can be improved by not using Git's recursive cloning anymore. I've assumed that the current implementation just was taking a short cut, but that longer term plan was to do exactly that. This basically means one would want to list only the (direct) submodules of a repository, e.g. `getNested()` would be changed. So, the case to list submodules without initializing them becomes the desired use case. 

So, I fear we're moving into the opposite direction here. @mnonnenmacher inupt would be interresting here as he has written the provenance resolver.

But going back from future to the code-as is, wouldn't the provenance resolver need to know if any submodule could not be intialized? (So that it can add the corresponding OrtIssue to the scan result)

Given these two, I'm still not convinced yet that the changes are desired.

Reply:
> Well, the downloader should resolve all submodules correctly if they are properly configured.
In case the project is not properly configured, we don't need this change at all because the project can simply be fixed. The issue arises when dependencies such as protobuf are not properly configured and the scanner crashes.

Reply:
> The issue arises when dependencies such as protobuf are not properly configured and the scanner crashes.

We should probably align on what we call "properly configured" WRT submodules: To me, a repo with uninitialized submodules is still properly configured. But a repo with submodules that point to the parent repo is not properly configured.

Reply:
Ok, let's use your definition

Reply:
is this conversation resolved?

Reply:
From my side, yes.

Reply:
Can we then resolve this conversation and merge the change? @fviernau

Reply:
@fviernau we are waiting on you

Reply:
@tsteenbe mentioned in the community meeting that @fviernau is sick. I guess it's ok to dismiss this review and eventually do a fixup.